### ISSUE TITLE
Integration with Keystone

### ISSUE THREAD

**Issue by VykintasZaboras** [buunguyen/mongoose-deep-populate#18]
I'm doing:
```
Post.paginate({
	page: req.query.page || 1,
	perPage: 50,
	maxPages: 10
}).populate('comments','author content date')
.deepPopulate('author.name')
```

and i included the plugin like so:
```
var deepPopulate = require('mongoose-deep-populate');
Post.schema.plugin(deepPopulate);
Post.register();
```

But i still get the error:
```
TypeError: undefined is not a function
```

Anyone got it working with keystone?

**Comment by buunguyen** [comment]
I answered this [here](https://github.com/buunguyen/mongoose-deep-populate/issues/1#issuecomment-70917678). You might want to send a question to KeystoneJS instead as I don't know how they wrap Mongoose.

**Comment by AndrewFinlay** [comment]
Hi there,

I was having this problem recently and made a few discoveries.

Vykintas, your setup appears to be very similar to mine and should work, although you can put your 1st level population requests into the deepPopulate call and specify your projection in the Post.schema.plugin call.  

I found that when 'deepPopulate' is installed it adds the latest version of mongoose as a dependency to your project if it isn't already specified, currently version 4.x but it can work with earlier releases.  Keystone also has mongoose as a dependency and uses version 3.8.x.  The two aren't compatible and when you use the plugin call, 'deepPopulate' doesn't end up where it's supposed to be and will fail when called.

To fix this simply add mongoose as a dependency to your package.json file at a version to match that found in Keystone (currently 3.8.x).

BTW, Thanks for a great module BN.

**Comment by buunguyen** [comment]
@AndrewFinlay great digging, thanks. This makes a lot of sense. If you can submit a PR for the README to highlight this, it's much appreciated.

**Comment by razb** [comment]
I must still be missing something. I 've verified that both the keystone and mongoose-deep-populate package.json files refrence the same mongoose version:

<pre>
  "peerDependencies": {
    "mongoose": "~3.8.24"
  },
</pre>

and have verified that the plugin is registered as a method for the schema:

<pre>
{ deepPopulate: [Function],
  getRelated: [Function],
  populateRelated: [Function],
  getUpdateHandler: [Function] }
</pre>

Yet I still get:

<pre>
TypeError: undefined is not a function
</pre>

Any suggestions?

**Comment by buunguyen** [comment]
The key thing is that both this plugin and Keystone must work on the same instance of `mongoose`. From my cursory look at Keystone, it looks like Keystone will use its own instance due to the declaration of mongoose as `dependency` (instead of `peerDependency`). So when Keystone and this plugin use `require` to load mongoose, each will have a different instance.

There are 2 ways to fix this:

1. See if you can pass the mongoose instance "required" from your app to Keystone to use. I'm not sure if Keystone has an API for this. If it does, then the code might look something like this:
```js
var mongoose = require('mongoose')
var keystone = require('keystone')
keystone.set('mongoose', mongoose)
```

2. Pass the mongoose instance used by Keystone to mongoose-deep-populate. You need to get code from the [experimental 2.0 branch](https://github.com/buunguyen/mongoose-deep-populate/tree/v2.0.0).
```js
var mongoose = mongoose_instance_used_by_keystone
var deepPopulate = require('mongoose-deep-populate')(mongoose)
```

**Comment by razb** [comment]
Thank you!

I used the second option:

<pre>
var mongoose = (keystone.mongoose)
var deepPopulate = require('mongoose-deep-populate')(mongoose);
</pre>

Works perfectly

**Comment by AndrewFinlay** [comment]
Modify the readme and create a pr?.

**Comment by buunguyen** [comment]
@AndrewFinlay already in the experimental branch.

**Comment by buunguyen** [comment]
Released 2.0.0.

**Comment by robksawyer** [comment]
@buunguyen I'm a bit unclear, have you gotten this working? It's not working for me using the methods described above.
